### PR TITLE
Created parser

### PR DESCRIPTION
**Pull request by antonromanov1**
Parse constructions below:

* Functions with parameters and return types,
* Variable declarations, assigns,
* Blocks of code,
* True, false and integer literals, minuses,
* Binary arithmetic operations, shifts,
* Conditional branches and conditional cycles,
* Checks on relations, equality,
* Calls

Written tests on lexing and parsing.

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -0,0 +1,997 @@
...
+            Token::LBrace => self.parse_block(),
+            Token::While => self.parse_while(),
+            Token::If => self.parse_if(),
+            _ => {
+                found(format!("statement can't start with '{}'", t.to_string()));
+            }
+        }
+    }
+
+    fn parse_func(&mut self, global: Env) -> Func {
```
Comment (zinoviy23) [inline src/parser.rs:543]:
```suggestion
    fn parse_func(&mut self, global: Env) -> ParseResult<Func> {
```

File: src/parser.rs [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+        self.go_next_token();
+        let func_name: String = self.consume_id();
+        self.expect(&Token::LParent);
+        let mut func_params: Vec<Node> = Vec::new();
+
+        loop {
+            if self.consume(&Token::RParent) {
+                break;
+            }
+            func_params.push(self.define_param());
```
Comment (zinoviy23) [inline src/parser.rs:555]:
```suggestion
            func_params.push(self.define_param()?);
```

File: src/parser.rs [hunk 3]
```diff
@@ -0,0 +1,997 @@
...
+            self.consume_typename();
+        }
+
+        let func_stmts: Vec<Node> = self.compound_stmt();
+        Func {
+            name: func_name,
+            params: func_params,
+            stmts: func_stmts,
+            env: self.cur_env.clone(),
+        }
```
Comment (zinoviy23) [inline src/parser.rs:572]:
```suggestion
        Ok(Func {
            name: func_name,
            params: func_params,
            stmts: func_stmts,
            env: self.cur_env.clone(),
        })
```

File: tests/parser_test.rs [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+        let num2 = 2 >> 1;
+    }
+    "
+    .to_string();
+
+    // Parse source into the AST nodes
+    let funcs = parse(source);
+
+    // Create expected nodes
+    let lshift = Node::LShift(Box::new(Node::Integer(1)), Box::new(Node::Integer(2)));
```
[the remark this answers is not in the document]
Comment (antonromanov1) [reply tests/parser_test.rs:252]: I created the issue, #2.

File: src/parser.rs [hunk 5]
```diff
@@ -0,0 +1,978 @@
...
+                break;
+            }
+        }
+
+        Ok(lhs)
+    }
+
+    fn add_sub(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.mul_div()?;
+        self.check_invalid(&lhs)?;
```
Comment (zinoviy23) [inline src/parser.rs:754]: I expect, that after calling `check_invalid` `lhs` on the next line will be invalid. Maybe, `ensure_valid` will solve the ambiguity.

File: src/parser.rs [hunk 6]
```diff
@@ -0,0 +1,978 @@
...
+                let t: Token = self.get_token();
+                match t {
+                    Token::LParent => {
+                        self.expect(&Token::LParent)?;
+                        let mut args: Vec<Node> = Vec::new();
+
+                        loop {
+                            if self.consume(&Token::RParent) {
+                                break;
+                            }
```
Comment (zinoviy23) [inline src/parser.rs:839]:
```suggestion
                        while !self.consume(&Token::RParent) {
```

File: src/parser.rs [hunk 7]
```diff
@@ -0,0 +1,978 @@
...
+
+    // Parse block of statements (begining with opening curly brace and ending with the closing one)
+    // and return vector of the nodes
+    fn compound_stmt(&mut self) -> ParseResult<Vec<Node>> {
+        let mut stmts: Vec<Node> = Vec::new();
+        self.expect(&Token::LBrace)?;
+        loop {
+            if self.consume(&Token::RBrace) {
+                break;
+            }
```
Comment (zinoviy23) [inline src/parser.rs:866]:
```suggestion
        while !self.consume(&Token::RBrace) {
```

File: src/parser.rs [hunk 8]
```diff
@@ -0,0 +1,978 @@
...
+        Ok(lhs)
+    }
+
+    fn mul_div(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.unary()?;
+        self.check_invalid(&lhs)?;
+        loop {
+            if !self.check_vec(vec![Token::Star, Token::Slash, Token::Percent]) {
+                break;
+            }
```
Comment (zinoviy23) [inline src/parser.rs:777]:
```suggestion
        while self.check_vec(vec![Token::Star, Token::Slash, Token::Percent]) {
```

File: src/parser.rs [hunk 9]
```diff
@@ -0,0 +1,978 @@
...
+        Ok(lhs)
+    }
+
+    fn add_sub(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.mul_div()?;
+        self.check_invalid(&lhs)?;
+        loop {
+            if !self.check_vec(vec![Token::Plus, Token::Minus]) {
+                break;
+            }
```
Comment (zinoviy23) [inline src/parser.rs:758]:
```suggestion
        while self.check_vec(vec![Token::Plus, Token::Minus]) {
```

File: src/parser.rs [hunk 10]
```diff
@@ -0,0 +1,978 @@
...
+    }
+
+    fn relation(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.shift()?;
+        self.check_invalid(&lhs)?;
+
+        loop {
+            if !self.check_vec(vec![Token::Lt, Token::Gt, Token::Le, Token::Ge]) {
+                break;
+            }
```
Comment (zinoviy23) [inline src/parser.rs:715]:
```suggestion
        while self.check_vec(vec![Token::Lt, Token::Gt, Token::Le, Token::Ge]) {
```

File: src/parser.rs [hunk 11]
```diff
@@ -0,0 +1,978 @@
...
+    }
+
+    fn equal(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.relation()?;
+        self.check_invalid(&lhs)?;
+
+        loop {
+            if !self.check_vec(vec![Token::Eq, Token::Ne]) {
+                break;
+            }
```
Comment (zinoviy23) [inline src/parser.rs:694]:
```suggestion
        while self.check_vec(vec![Token::Eq, Token::Ne]) {
```

File: src/parser.rs [hunk 12]
```diff
@@ -0,0 +1,978 @@
...
+        self.cur_env.prev = Some(Box::new(global));
+        self.go_next_token();
+        let func_name: String = self.consume_id()?;
+        self.expect(&Token::LParent)?;
+        let mut func_params: Vec<Node> = Vec::new();
+
+        loop {
+            if self.consume(&Token::RParent) {
+                break;
+            }
```
Comment (zinoviy23) [inline src/parser.rs:551]:
```suggestion
        while !self.consume(&Token::RParent) {
```

File: src/parser.rs [hunk 13]
```diff
@@ -0,0 +1,978 @@
...
+        Ok(())
+    }
+
+    fn check_vec(&self, tks: Vec<Token>) -> bool {
+        for t in tks.iter() {
+            if self.cur_token() == t {
+                return true;
+            }
+        }
+        false
```
Comment (zinoviy23) [inline src/parser.rs:936]:
```suggestion
        tks.iter().any(|t| t == self.cur_token())
```

**Review by zinoviy23 — APPROVED** [review]

**Pull request merged**